### PR TITLE
Other pandoc formats

### PR DESCRIPTION
**Pull request by johndgiese**
# Author's Checklist

Before assigning reviewers, step through the following checklist. Check each item once it's completed. If an item is skipped, write an explanation below the item.

- [ ] **Construction:** Review the code diff and confirm there's no dead code, debug code, or unnecessary comments.
- [ ] **Requirements:** Review the requirements in the issue for this PR and confirm they're all implemented.
- [ ] **Architecture:** Confirm that the changes to the code are consistent with the existing architecture.
- [ ] **SOUP/OTS Software:** Any new dependencies have appropriate licenses and are documented or pinned in requirements files.
- [ ] **Unit Test:** Atomic unit tests have been added, if appropriate.
- [ ] **End-to-End Tests:** End-to-end tests have been extended or updated, if appropriate.
- [ ] **Low-level Documentation:** Comments or doc strings for affected code have been updated.
- [ ] **High-Level Documentation:** The README is updated, as appropriate.
- [ ] **Change Log:** New features or bug fixes have been added to the change log in the README.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -344,6 +371,13 @@ Here are some features we're planning to add in the future:
 
 ## Changelog
 
+### v0.9.0
+
+- Replace the `filename_property` configuration option with the more generic `filename_template`.
+- Make it so we can render apges into any pandoc format
```
Comment (dwhensley) [inline README.md:377]: typo: `apges` -> `pages`

File: README.md [hunk 2]
```diff
@@ -243,11 +244,37 @@ Most of the Notion blocks can generate their pandoc AST from _only_ their own da
...
-### Render
+### Jinja Render Page
 
-When CodeBlocks are captioned "{template}", the block is rendered as jinja using yaml data previously cached by n2y.
+CodeBlocks whose caption begins with "{jinja=pandocformat}" will be rendered using [Jinja](https://jinja.palletsprojects.com/en/3.1.x/templates/) into text and then read into the AST using the specified pandoc input format. Note that, other than the special "plain" format, the pandocformat must be available both for reading and writing.
 
-The jinja context has a few special filters available to it. See the code for details.
+Any databases that are [mentioned](https://www.notion.so/help/comments-mentions-and-reminders) in the codeblock's caption will be made available in the jinja render context within the `databases` dictionary.
+
+For example, if you have a code block with the caption "{jinja=html} @People:
```
Comment (dwhensley) [inline README.md:253]: Missing terminal `"` and incomplete sentence.

File: n2y/config.py [hunk 3]
```diff
@@ -127,6 +135,18 @@ def _validate_config_item(config_item):
     return True
 
 
+def _valid_filename_template(filename_template):
+    if not isinstance(filename_template, str):
+        logger.error("filename_template must be a string")
+        return False
+    try:
+        string.Formatter().parse(filename_template)
```
Comment (dwhensley) [inline n2y/config.py:143]: Under what conditions will this fail/are you looking for?

File: n2y/notion.py [hunk 4]
```diff
@@ -24,6 +24,9 @@
     UseNextClass, is_api_error_code, APIErrorCode
 )
 
+
+# TODO: Rename this file `client.py`
```
Comment (dwhensley) [inline n2y/notion.py:28]: Is this TODO to be addressed in this PR?

File: n2y/plugins/jinjarenderpage.py [hunk 5]
```diff
@@ -182,7 +183,13 @@ def __init__(self, client, notion_data, page, get_children=True):
             # defaults and it could provide a mechanism to access the page
             # content from within the Jinja templates, which isn't possible
             # right now.
-            self.databases[database.title.to_plain_text()] = database_to_yaml(
+            database_name = database.title.to_plain_text()
+            if database_name in self.databases:
```
Comment (dwhensley) [inline n2y/plugins/jinjarenderpage.py:187]: Should this be a hard error?

File: n2y/utils.py [hunk 6]
```diff
@@ -93,6 +94,51 @@ def pandoc_write_or_log_errors(pandoc_ast, format, options):
...
+    elif isinstance(data, bool):
+        return MetaBool(data)
+    elif data is None:
+        return MetaString("")
+    elif isinstance(data, numbers.Number):
+        return MetaString(str(data))
+    elif isinstance(data, list):
+        return MetaList([yaml_to_meta_value(item) for item in data])
+    elif isinstance(data, dict):
+        return MetaMap({key: yaml_to_meta_value(value) for key, value in data.items()})
```
Comment (dwhensley) [inline n2y/utils.py:117]: `if`-`else` branching is non-exhaustive. Should a terminal `else` clause be added to at least log a warning on top of the current behavior of implicitly returning `None`?

File: n2y/utils.py [hunk 7]
```diff
@@ -93,6 +94,51 @@ def pandoc_write_or_log_errors(pandoc_ast, format, options):
...
+
+def yaml_map_to_meta(data):
+    assert isinstance(data, dict)
+    assert all(isinstance(k, str) for k in data.keys())
+    return Meta({k: yaml_to_meta_value(v) for k, v in data.items()})
+
+
+def pandoc_format_to_file_extension(format):
+    # split on '-' or '+' to handle formats like 'markdown+raw_tex'
+    base_type = re.split(r'[-+]', format)[0]
```
Comment (dwhensley) [inline n2y/utils.py:128]: Do you not ever need to consider the post `+` or `-` extension type?

**Review by dwhensley — COMMENTED** [review]
I performed a quick, high-level review. I left just a few comments and questions.

File: n2y/config.py (hunk 3 above)
Comment (sHermanGriffiths) [reply n2y/config.py:143]: I think he meant to check that filename_template is a format string and I have adjusted this function to properly reflect that.

File: n2y/notion.py (hunk 4 above)
Comment (sHermanGriffiths) [reply n2y/notion.py:28]: No it is not.

File: n2y/plugins/jinjarenderpage.py (hunk 5 above)
Comment (sHermanGriffiths) [reply n2y/plugins/jinjarenderpage.py:187]: I actually agree that it should be.

Comment (johndgiese) [reply n2y/plugins/jinjarenderpage.py:187]: Great! I'm assuming you'll implement these changes :)

File: n2y/utils.py (hunk 6 above)
Comment (sHermanGriffiths) [reply n2y/utils.py:117]: Good Idea, I think so.

File: n2y/utils.py (hunk 7 above)
Comment (sHermanGriffiths) [reply n2y/utils.py:128]: I don't believe so, as I have yet to see an extension that so drastically changes the format that we'd need to adjust the file type to it.

**Pull request merged**